### PR TITLE
Introduce Cura specific ContainerStack subclasses

### PR DESCRIPTION
**Pull request by awhiemstra**
This introduces a CuraContainerStack class with two subclasses, ExtruderStack and GlobalStack, along with code to create these and to upgrade the old ContainerStack instances to the new Extruder/Global stack classes.

This is the ground work for simplifying and cleaning up a bunch of the setting code in Cura. The amount of changes to the rest of the code have been intentionally left as small as possible, to prevent serious regressions. Once we land this code, we can start deprecating and replacing code relying on the old code.

This PR depends on https://github.com/Ultimaker/Uranium/pull/231

Note that CI reports test failures because of the missing decorator and does not correctly detect the dependency on the Uranium PR.
Additional note: There are a few unit tests in TestCuraContainerRegistry that fail. @Ghostkeeper is still working on those.

### CODE REVIEW

File: tests/Settings/TestGlobalStack.py [hunk 1]
```diff
@@ -0,0 +1,546 @@
+# Copyright (c) 2017 Ultimaker B.V.
```
Comment (jackha) [inline tests/Settings/TestGlobalStack.py]: [already discussed, good for now]. It makes more sense to have tests for CuraContainerStack too, that prevents test duplication.

File: tests/Settings/TestGlobalStack.py [hunk 2]
```diff
@@ -0,0 +1,546 @@
...
+# Cura is released under the terms of the AGPLv3 or higher.
+
```
Comment (jackha) [inline tests/Settings/TestGlobalStack.py]: I'm missing tests for findDefault* methods.

File: cura/Settings/CuraContainerRegistry.py [hunk 3]
```diff
@@ -284,3 +299,27 @@ def _machineHasOwnQualities(self):
...
+    ##  Convert an "old-style" pure ContainerStack to either an Extruder or Global stack.
+    def _convertContainerStack(self, container):
+        assert type(container) == ContainerStack
+
+        container_type = container.getMetaDataEntry("type")
+        if container_type not in ("extruder_train", "machine"):
+            # It is not an extruder or machine, so do nothing with the stack
+            return container
+
+        new_stack = None
```
Comment (nallath) [inline cura/Settings/CuraContainerRegistry.py]: This line doesn't do anything. (minor issue)

File: cura/Settings/CuraContainerStack.py [hunk 4]
```diff
@@ -0,0 +1,616 @@
...
+    #   \param new_user_changes The new user changes container. It is expected to have a "type" metadata entry with the value "user".
+    def setUserChanges(self, new_user_changes: InstanceContainer) -> None:
+        self.replaceContainer(_ContainerIndexes.UserChanges, new_user_changes)
+
+    ##  Get the user changes container.
+    #
+    #   \return The user changes container. Should always be a valid container, but can be equal to the empty InstanceContainer.
+    @pyqtProperty(InstanceContainer, fset = setUserChanges, notify = pyqtContainersChanged)
+    def userChanges(self) -> InstanceContainer:
+        return self._containers[_ContainerIndexes.UserChanges]
```
Comment (nallath) [inline cura/Settings/CuraContainerStack.py]: userChanges, as it's a pyqtProperty.

File: cura/Settings/CuraContainerStack.py [hunk 5]
```diff
@@ -0,0 +1,616 @@
...
+        self.replaceContainer(_ContainerIndexes.QualityChanges, new_quality_changes)
+
+    ##  Set the quality changes container by an ID.
+    #
+    #   This will search for the specified container and set it. If no container was found, an error will be raised.
+    #
+    #   \param new_quality_changes_id The ID of the new quality changes container.
+    #
+    #   \throws Exceptions.InvalidContainerError Raised when no container could be found with the specified ID.
+    def setQualityChangesById(self, new_quality_changes_id: str) -> None:
```
Comment (nallath) [inline cura/Settings/CuraContainerStack.py]: Good point. I'd say yes. At the very least some logging.

Comment (nallath) [inline cura/Settings/CuraContainerStack.py]: Nevermind. The code already does this. The replaceContainer checks if the type is correct (and setQualityChanges uses replace container)

File: cura/Settings/CuraContainerStack.py [hunk 6]
```diff
@@ -0,0 +1,616 @@
...
+    #   no container with the correct type be found, it will use the empty container.
+    #
+    #   \throws InvalidContainerStackError Raised when no definition can be found for the stack.
+    @override(ContainerStack)
+    def deserialize(self, contents: str) -> None:
+        super().deserialize(contents)
+
+        new_containers = self._containers.copy()
+        while(len(new_containers) < len(_ContainerIndexes.IndexTypeMap)):
+            new_containers.append(self._empty_instance_container)
```
Comment (nallath) [inline cura/Settings/CuraContainerStack.py]: What do you mean?

File: cura/Settings/CuraContainerStack.py [hunk 7]
```diff
@@ -0,0 +1,606 @@
...
+
+        preferred_variant_id = definition.getMetaDataEntry("preferred_variant")
+        if preferred_variant_id:
+            preferred_variants = ContainerRegistry.getInstance().findInstanceContainers(id = preferred_variant_id, definition = definition_id, type = "variant")
+            if preferred_variants:
+                variant = preferred_variants[0]
+            else:
+                Logger.log("w", "The preferred variant \"{variant}\" of stack {stack} does not exist or is not a variant.", variant = preferred_variant_id, stack = self.id)
+                # And leave it at the default variant.
+
```
Comment (nallath) [inline cura/Settings/CuraContainerStack.py]: Isnt the following a bit cleaner:
```
if not variant:
    Logger.log("w", "Could not find a valid default variant for stack {stack}", stack = self.id)
return variant
```

File: cura/Settings/CuraContainerStack.py [hunk 8]
```diff
@@ -0,0 +1,606 @@
...
+            containers = registry.findInstanceContainers(**search_criteria)
+            if containers:
+                return containers[0]
+
+        return None
+
+    ## protected:
+
+    # Helper to make sure we emit a PyQt signal on container changes.
+    def _onContainersChanged(self, container: Any) -> None:
```
Comment (nallath) [inline cura/Settings/CuraContainerStack.py]: Is it really any object that should work?

File: cura/Settings/CuraStackBuilder.py [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+from UM.Settings.DefinitionContainer import DefinitionContainer
+from UM.Settings.InstanceContainer import InstanceContainer
+from UM.Settings.ContainerRegistry import ContainerRegistry
+
+from .GlobalStack import GlobalStack
+from .ExtruderStack import ExtruderStack
+from .CuraContainerStack import CuraContainerStack
+
+##  Contains helper functions to create new machines.
+class CuraStackBuilder:
```
Comment (nallath) [inline cura/Settings/CuraStackBuilder.py]: Yeah, this will save us sooo much time <3

File: cura/Settings/ExtruderStack.py [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+    def __init__(self, container_id, *args, **kwargs):
+        super().__init__(container_id, *args, **kwargs)
+
+        self.addMetaDataEntry("type", "extruder_train") # For backward compatibility
+
+    ##  Overridden from ContainerStack
+    #
+    #   This will set the next stack and ensure that we register this stack as an extruder.
+    @override(ContainerStack)
+    def setNextStack(self, stack: ContainerStack) -> None:
```
Comment (nallath) [inline cura/Settings/ExtruderStack.py]: I don't really see why. I think parent stack is more confusing.

File: cura/Settings/GlobalStack.py [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+from UM.Settings.ContainerStack import ContainerStack, InvalidContainerStackError
+from UM.Settings.InstanceContainer import InstanceContainer
+from UM.Settings.SettingInstance import InstanceState
+from UM.Settings.DefinitionContainer import DefinitionContainer
+from UM.Settings.ContainerRegistry import ContainerRegistry
+from UM.Settings.Interfaces import ContainerInterface
+
+from . import Exceptions
+from .CuraContainerStack import CuraContainerStack
+
```
Comment (nallath) [inline cura/Settings/GlobalStack.py]: pep sez: more blank lines

**Comment by fieldOfView** [comment]
And so the counter is reset:
dayssinceaccident

File: cura/Settings/CuraContainerStack.py (hunk 6 above)
Comment (Ghostkeeper) [inline cura/Settings/CuraContainerStack.py]: The extruder stack doesn't need to list all containers in the file. The rest are filled with empty containers. Some test cases actually test this behaviour.

**Pull request merged**